File: aochelpers/input.go
package aochelpers

import (
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"os"
	"os/exec"
	"os/user"
	"path/filepath"
	"strings"
)

func getCacheDirectory() string {
	user, err := user.Current()
	if err != nil {
		panic(err)
	}
	return fmt.Sprintf("%s\\.advent-of-code", user.HomeDir)
}

func fileExists(filename string) bool {
	info, err := os.Stat(filename)
	if os.IsNotExist(err) {
		return false
	}
	return !info.IsDir()
}

func readFileIntoString(filePath string) string {
	bytes, err := ioutil.ReadFile(filePath)
	if err != nil {
		panic(err)
	}
	return strings.TrimRight(string(bytes), "\n")
}

func getCookieCacheFile() string {
	return fmt.Sprintf("%s\\session-cookie.txt", getCacheDirectory())
}

func forgetSessionCookie() {
	os.Remove(getCookieCacheFile())
}

func getSessionCookie() string {
	cookieCacheFile := getCookieCacheFile()
	if fileExists(cookieCacheFile) {
		return readFileIntoString(cookieCacheFile)
	}
	newCookieQuery := `A new Advent of Code session cookie is needed. Please do the following:
1) Open this in chrome: chrome://settings/cookies/detail?site=adventofcode.com
2) Look at the session cookie and copy the value/content
3) Replace this file with that cookie`
	ioutil.WriteFile(cookieCacheFile, []byte(newCookieQuery), os.ModePerm)
	exec.Command("notepad", cookieCacheFile).Run()
	return readFileIntoString(cookieCacheFile)
}

func downloadFile(url string, filename string, cookies ...http.Cookie) {
	client := new(http.Client)

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		panic(err)
	}

	for _, c := range cookies {
		heapCookie := new(http.Cookie)
		*heapCookie = c
		req.AddCookie(heapCookie)
	}

	resp, err := client.Do(req)
	if err != nil {
		panic(err)
	}
	defer resp.Body.Close()

	// Create the file
	out, err := os.Create(filename)
	if err != nil {
		panic(err)
	}
	defer out.Close()

	// Write the body to file
	_, err = io.Copy(out, resp.Body)
	if err != nil {
		panic(err)
	}
}

func getInputCacheFile(year int, day int) string {
	return fmt.Sprintf("%s\\%d\\day-%d.txt", getCacheDirectory(), year, day)
}

// GetInput returns the input for a given problem (by year/day)
// It downloads these inputs from the server on-demand, intelligently caching them on disk
// to avoid pinging the server repeatedly. It also has rudimentary handling to detect
// login issues and premature inputs
func GetInput(year int, day int) string {
	notLoggedInFile := "Puzzle inputs differ by user.  Please log in to get your puzzle input."
	tooEarlyFile := "Please don't repeatedly request this endpoint before it unlocks! The calendar countdown is synchronized with the server time; the link will be enabled on the calendar the instant this puzzle becomes available."

	inputFilePath := getInputCacheFile(year, day)
	os.MkdirAll(filepath.Dir(inputFilePath), os.ModePerm)
	if fileExists(inputFilePath) {
		contents := readFileIntoString(inputFilePath)
		if contents != notLoggedInFile && contents != tooEarlyFile {
			// The contents are good! (I think)
			return contents
		}
		os.Remove(inputFilePath)
	}

	url := fmt.Sprintf("https://adventofcode.com/%d/day/%d/input", year, day)
	sessionCookie := http.Cookie{Name: "session", Value: getSessionCookie()}

	downloadFile(url, inputFilePath, sessionCookie)

	contents := readFileIntoString(inputFilePath)
	if contents != notLoggedInFile {
		// The contents are good! (I think)
		return contents
	}

	// The session cookie may be invalid?
	os.Remove(inputFilePath)
	forgetSessionCookie()
	sessionCookie.Value = getSessionCookie()

	// Last try
	downloadFile(url, inputFilePath, sessionCookie)
	return readFileIntoString(inputFilePath)
}

// ForgetInput removes the input from the cache (effectively forcing a re-download on the next attempt)
func ForgetInput(year int, day int) {
	os.Remove(getInputCacheFile(year, day))
}

// ClearInputCache clears the entire input cache (except for the session cookie)
func ClearInputCache() {
	cacheDir := getCacheDirectory()

	// Make sure it exists, it's just easier this way
	os.MkdirAll(cacheDir, os.ModePerm)

	cookieFile := filepath.Base(getCookieCacheFile())

	items, _ := ioutil.ReadDir(cacheDir)
	for _, i := range items {
		if cookieFile == i.Name() {
			continue
		}
		toDelete := filepath.Join(cacheDir, i.Name())
		os.RemoveAll(toDelete)
	}
}

File: helpers/input.py
import os
import pathlib
import requests
import shutil
import subprocess

_s = requests.Session()

def _get_cache_directory():
    return pathlib.Path.home() / '.advent-of-code'

def _read_file_as_string(path):
    with open(path) as f:
        return f.read().rstrip('\n')

def _get_cookie_cache_file():
    return _get_cache_directory() / 'session-cookie.txt'

def _forget_session_cookie():
    _get_cookie_cache_file().unlink()
    del _s.cookies['session']

def _load_session_cookie():
    cookie_cache_file = _get_cookie_cache_file()

    if not cookie_cache_file.exists():
        cookie_cache_file.write_text('''A new Advent of Code session cookie is needed. Please do the following:
1) Open this in chrome: chrome://settings/cookies/detail?site=adventofcode.com
2) Look at the session cookie and copy the value/content
3) Replace this file with that cookie''')

        subprocess.check_call(['notepad', cookie_cache_file])

    _s.cookies['session'] = cookie_cache_file.read_text().rstrip('\n')

def _download_file(url, file_path):
    r = _s.get(url)
    file_path.write_bytes(r.content)

def _get_input_cache_file(year, day):
    return _get_cache_directory() / str(year) / f'day-{day}.txt'

def get_input(year, day):
    '''
get_input returns the input for a given problem (by year/day)
It downloads these inputs from the server on-demand, intelligently caching them on disk
to avoid pinging the server repeatedly. It also has rudimentary handling to detect
login issues and premature inputs
    '''
    notLoggedInFile = 'Puzzle inputs differ by user.  Please log in to get your puzzle input.'
    tooEarlyFile = 'Please don\'t repeatedly request this endpoint before it unlocks! The calendar countdown is synchronized with the server time; the link will be enabled on the calendar the instant this puzzle becomes available.'

    input_file_path = _get_input_cache_file(year, day)

    if input_file_path.exists():
        contents = input_file_path.read_text().rstrip('\n')
        if contents != notLoggedInFile and contents != tooEarlyFile:
            # The contents are good! (I think)
            return contents

        input_file_path.unlink()
    os.makedirs(input_file_path.parent, exist_ok=True)

    url = f'https://adventofcode.com/{year}/day/{day}/input'
    _load_session_cookie()

    _download_file(url, input_file_path)

    contents = input_file_path.read_text().rstrip('\n')
    if contents != notLoggedInFile:
        # The contents are good! (I think)
        return contents

    # The session cookie may be invalid?
    input_file_path.unlink()
    _forget_session_cookie()
    _load_session_cookie()

    # Last try
    _download_file(url, input_file_path)
    return input_file_path.read_text().rstrip('\n')

def forget_input(year, day):
    '''forget_input removes the input from the cache (effectively forcing a re-download on the next attempt)'''
    _get_input_cache_file(year, day).unlink()

def clear_input_cache():
    '''clear_input_cache clears the entire input cache (except for the session cookie)'''
    cache_dir = _get_cache_directory()
    cookie_file = _get_cookie_cache_file()

    os.makedirs(cache_dir, exist_ok=True)

    for child in list(cache_dir.iterdir()):
        if child == cookie_file:
            continue
        shutil.rmtree(child)
